tika-parsers/src/main/java/org/apache/tika/parser/html/BoilerpipeContentHandler.java

```diff
@@ -230,14 +229,20 @@ public void endDocument() throws SAXException {
                     case CONTINUE:
                         // Now emit characters that are valid. Note that boilerpipe pre-increments the character index, so
                         // we have to follow suit.
-                        for (char[] chars : element.getCharacters()) {
+                        for (int i = 0; i < element.getCharacters().size(); i++) {
+                            char[] chars = element.getCharacters().get(i);
                             curCharsIndex++;
+                            boolean isValidCharacterRun = validCharacterRuns.get(curCharsIndex);
 
-                            if (validCharacterRuns.get(curCharsIndex)) {
+                            // https://issues.apache.org/jira/projects/TIKA/issues/TIKA-2683
+                            // Allow exempted characters to be written
+                            if (isValidCharacterRun ||
+                                    (chars.length == 1 && whitelistCharSet.contains(chars[0]))) {
                                 delegate.characters(chars, 0, chars.length);
 
                                 // https://issues.apache.org/jira/browse/TIKA-961
-                                if (!Character.isWhitespace(chars[chars.length - 1])) {
+                                if (isValidCharacterRun && i == element.getCharacters().size() - 1
```

Comment:
I'd move this block up one level - I don't believe it depends on the previous test (`isValidCharacterRun || (chars.length == 1...`) evaluating to true, right?

Reply:
That's correct. I will move this one level up. 👍